### PR TITLE
feat: add matrix-auth integration

### PR DESCRIPTION
**Pull request by amandahla**
### Overview

Add matrix-auth integration.

- Plugins:
  - Requires information only (homeserver and shared secret) 
  - Example: Maubot
- Application services
  - Requires information and also provides secret to be registered in Synapse.
  - Example: IRC Bridge. See the [docs](https://matrix-org.github.io/matrix-appservice-irc/latest/bridge_setup.html#4-registration) for more details about it.
 
Note:
Twisted version is pinned due to this issue:
https://github.com/element-hq/synapse/issues/17882

#### Todo

- [x] Plugins
- [x] Application Services
- [x] Unit tests
- [ ] Integration tests
   

### Rationale

Integrate Synapse with plugins and application services

### Juju Events Changes

Matrix auth relation changed and departed.
Why not matrix_auth_request_processed?
The `on = RedisRelationCharmEvents()` prevents that. :-(

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Manging Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

**Review by nrobinaubertin — APPROVED** [review]

**Comment by amandahla** [comment]
Integration test failing due
https://github.com/element-hq/synapse/issues/17882

File: src/matrix_auth_observer.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        """
+        return self._charm
+
+    def update_matrix_auth_integration(self, charm_state: CharmState) -> None:
+        """Update matrix auth integration relation data.
+
+        Args:
+            charm_state: The charm state.
+        """
+        for relation in list(self._charm.model.relations["matrix-auth"]):
```
Comment (yanksyoon) [inline src/matrix_auth_observer.py:59]: Out of curiosity, is `list(...)` to guard against the relations being None?

**Review by Thanhphan1147 — DISMISSED** [review]

File: src/matrix_auth_observer.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            self.matrix_auth.update_relation_data(relation, provider_data)
+
+    def get_requirer_registration_secrets(self) -> Optional[List]:
+        """Get requirers registration secrets (application services).
+
+        Returns:
+            dict with filepath and content for creating the secret files.
+        """
+        registration_secrets = []
+        RegistrationSecret = namedtuple("RegistrationSecret", ["file_path", "value"])
```
Comment (yanksyoon) [inline src/matrix_auth_observer.py:74]: Would it help to have a separate class defined for it?
```suggestion
class RegistrationSecret(typing.NamedTuple):
    file_path: pathlib.Path
    value: str
```

File: src/synapse/workload.py [hunk 3]
```diff
@@ -404,3 +404,16 @@ def generate_mjolnir_config(container: ops.Container, access_token: str, room_id
...
+    """Create registration secrets files.
+
+    Args:
+        container: Container of the charm.
+        charm_state: Instance of CharmState.
+    """
+    container.exec(["rm", "-f", f"{SYNAPSE_CONFIG_DIR}/appservice-registration-*.yaml"])
+    if charm_state.registration_secrets:
+        for registration_secret in charm_state.registration_secrets:
+            registration_secret.file_path.write_text(registration_secret.value)
```
Comment (yanksyoon) [inline src/synapse/workload.py:419]: Should we define the encoding here?
```suggestion
            registration_secret.file_path.write_text(registration_secret.value, encoding="utf-8")
```

File: src/synapse/workload_configuration.py [hunk 4]
```diff
@@ -265,6 +265,27 @@ def enable_redis(current_yaml: dict, charm_state: CharmState) -> None:
...
+
+    Args:
+        current_yaml: current configuration.
+        charm_state: Instance of CharmState.
+
+    Raises:
+        WorkloadError: something went wrong enabling registration secrets.
+    """
+    try:
+        if charm_state.registration_secrets is None:
```
Comment (yanksyoon) [inline src/synapse/workload_configuration.py:279]: Could this be simplified to `if not charm_state.registration_secrets`? checking for None is usually for `null` type of checks where it is purposely left empty to carry meaning.

**Review by yanksyoon — DISMISSED** [review]
Very nicely commented code! LGTM w minor nits :)

File: src/matrix_auth_observer.py (hunk 1 above)
Comment (amandahla) [reply src/matrix_auth_observer.py:59]: Exactly :)

File: src/synapse/workload_configuration.py (hunk 4 above)
Comment (amandahla) [reply src/synapse/workload_configuration.py:279]: Most of the code we have checks like this so I'm afraid that changing this one would be confusing. But is a good point for a refactoring that can be done in a different PR, WDYT

**Review by Thanhphan1147 — APPROVED** [review]

**Review by yanksyoon — APPROVED** [review]
Thanks for the quick changes!

File: src/synapse/workload_configuration.py (hunk 4 above)
Comment (amandahla) [reply src/synapse/workload_configuration.py:279]: ISD-2686 created to address this comment.

**Pull request merged**